File: LeetCode/Top-Interview-150/HashMap/Word_Pattern.java
public class Solution 
{
    public boolean wordPattern(String pattern, String s) 
    {
        String[] words = s.split(" ");
        
        if (pattern.length() != words.length) 
        {
            return false;
        }
        
        Map<Character, String> charToWord = new HashMap<>();
        Map<String, Character> wordToChar = new HashMap<>();
        
        for (int i = 0; i < pattern.length(); i++) 
        {
            char ch = pattern.charAt(i);
            String word = words[i];
            
            if (charToWord.containsKey(ch) && !charToWord.get(ch).equals(word)) 
            {
                return false;
            }
            
            if (wordToChar.containsKey(word) && wordToChar.get(word) != ch) 
            {
                return false;
            }
            
            charToWord.put(ch, word);
            wordToChar.put(word, ch);
        }
        
        return true;
    }
    
    public static void main(String[] args) 
    {
        Solution sol = new Solution();

        // Test cases
        System.out.println(sol.wordPattern("abba", "dog cat cat dog"));  // Expected output: true
        System.out.println(sol.wordPattern("abba", "dog cat cat fish")); // Expected output: false
        System.out.println(sol.wordPattern("aaaa", "dog cat cat dog"));  // Expected output: false
    }
}

/*

Time Complexity:

- O(n)
- We iterate through the pattern and the words once

 */

File: LeetCode/Top-Interview-150/HashMap/Word_Pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        
        words = s.split()

        if len(pattern) != len(words):
            return False

        char_to_word = {}
        word_to_char = {}

        for char, word in zip(pattern, words):

            if (char in char_to_word and char_to_word[char]!=word) or \
               (word in word_to_char and word_to_char[word]!=char):

               return False
            
            char_to_word[char] = word
            word_to_char[word] = char

        return True

# Create an instance of the Solution class
sol = Solution()

# Test cases
print(sol.wordPattern("abba", "dog cat cat dog"))  # Expected output: True
print(sol.wordPattern("abba", "dog cat cat fish")) # Expected output: False
print(sol.wordPattern("aaaa", "dog cat cat dog"))  # Expected output: False

'''

Time Complexity:

- O(n)
- We iterate through the pattern and the words once

'''